### PR TITLE
`n_jobs` support details in docs

### PR DESCRIPTION
**Pull request by Alexsandruss**
## Description

---

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [ ] I have reviewed my changes thoroughly before submitting this pull request.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [ ] I have added a respective label(s) to PR if I have a permission for that.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

N/A

### CODE REVIEW

File: doc/sources/parallelism.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+.. WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+.. See the License for the specific language governing permissions and
+.. limitations under the License.
+
+.. include:: substitutions.rst
+.. _parallelism:
+
+#####################
+Parallelism Specifics
+#####################
```
Comment (david-cortes-intel) [inline doc/sources/parallelism.rst:20]:
```suggestion
Parallelism Specifics
=====================
```

File: doc/sources/parallelism.rst [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+|sklearnex| supports the `n_jobs <https://scikit-learn.org/stable/glossary.html#term-n_jobs>`_ parameter
+of the original |sklearn| with the following differences:
```
Comment (david-cortes-intel) [inline doc/sources/parallelism.rst:23]:
```suggestion
|sklearnex| supports the `n_jobs <https://scikit-learn.org/stable/glossary.html#term-n_jobs>`_ parameter
of the original |sklearn|, with the following differences:
```

File: doc/sources/parallelism.rst [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+When Scikit-learn's utilities with built-in parallelism are used (for example, `GridSearchCV` or `VotingClassifier`),
+|sklearnex| tries to determine the optimal number of threads per job using hints provided by `joblib`.
+If `n_jobs` is not specified for underlying estimator(s), |sklearnex| sets it to the number of available threads
+(usually the number of logical CPUs divided by `n_jobs` set for higher-level parallelized entities).
+
+Environment variables such as `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, `OPENBLAS_NUM_THREADS`, and others used by
+low-level parallelism libraries are recognized by `joblib` and therefore can be used as hints by |sklearnex|.
+
+To track the actual number of threads used by sklearnex's estimators,
```
Comment (david-cortes-intel) [inline doc/sources/parallelism.rst:45]: There's also the MKL debug variable, and now the oneDAL debug variable.

File: doc/sources/parallelism.rst [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+* |sklearnex| doesn't use `joblib` for parallelism in patched estimators and functions
+* The only low-level parallelism library used by |sklearnex| is oneTBB (through oneDAL)
+* The `threading` parallel backend of `joblib` is not supported by |sklearnex|
+
+The only exception is multiclass LogisticRegression, which uses `joblib` for parallelism across classes.
+
+|sklearnex| follows the same rules as |sklearn| for
+`the calculation of the 'n_jobs' parameter value <https://scikit-learn.org/stable/glossary.html#term-n_jobs>`_.
+
+When Scikit-learn's utilities with built-in parallelism are used (for example, `GridSearchCV` or `VotingClassifier`),
```
Comment (david-cortes-intel) [inline doc/sources/parallelism.rst:37]:
```suggestion
When Scikit-learn's utilities with built-in parallelism are used (for example, :obj:`sklearn.model_selection.GridSearchCV` or :obj:`sklearn.model_selection.VotingClassifier`),
```

File: doc/sources/parallelism.rst [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+|sklearnex| tries to determine the optimal number of threads per job using hints provided by `joblib`.
```
Comment (david-cortes-intel) [inline doc/sources/parallelism.rst:38]: I wasn't aware that such a system existed. Could you provide a link to the code where this happens?

File: doc/sources/parallelism.rst [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+|sklearnex| supports the `n_jobs <https://scikit-learn.org/stable/glossary.html#term-n_jobs>`_ parameter
+of the original |sklearn| with the following differences:
+
+* `n_jobs` parameter is supported for all estimators patched by |sklearnex|,
+  while |sklearn| enables it for selected estimators only
+* `n_jobs` estimator parameter sets the number of threads used by the underlying |oneDAL|
+* |sklearnex| doesn't use `joblib` for parallelism in patched estimators and functions
+* The only low-level parallelism library used by |sklearnex| is oneTBB (through oneDAL)
+* The `threading` parallel backend of `joblib` is not supported by |sklearnex|
```
Comment (david-cortes-intel) [inline doc/sources/parallelism.rst:30]: What does it mean "not supported by sklearnex"? What happens for example if you run an sklearn metaestimator (like `BaggingClassifier`) using that joblib backend with an sklearnex estimator inside?

**Comment by david-cortes-intel** [comment]
Thanks for adding these explanations. But it's still missing important pieces of information and leaves several questions unanswered:

* It's missing the threading part of MKL, the static linkage part, and how it interacts with environment variables, `n_jobs` parameter, `inner_max_num_threads` parameter, and threadpoolctl configurations.
* It doesn't mention how the threading works when put under a threadpoolctl context.
* The explanation is unclear about what ends up happening with the number of threads when using environment variables in addition to passing `n_jobs` as parameter.
* Could mention what happens with `n_jobs` in GPU mode.
* There's a difference in the threading configuration logic between daal4py and sklearnex, which this doc could also mention.
* It doesn't cover the part about some configurations being global, which is quite relevant when using python-based multi-threading.
* It could mention that the TBB threading doesn't automatically avoid nested parallelism when used in conjunction with OpenMP (which sklearn uses) and/or with joblib or python threads.
* Some estimators perform better when not using all threads - for example, linear regression is faster on LNL laptops when not using low-power E-cores. Perhaps could mention these sort of things here as they are relevant.

**Comment by icfaust** [comment]
@Alexsandruss make sure to merge main for latest CI checks on docs

**Comment by david-cortes-intel** [comment]
Closing in favor of https://github.com/uxlfoundation/scikit-learn-intelex/pull/2768
